### ISSUE TITLE
terraform-oci-oke module deployment fails randomly

### ISSUE THREAD

**Issue by bieksaz** [oracle-terraform-modules/terraform-oci-oke#225]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version and Provider Version

Terraform v0.12.29
terragrunt version v0.25.4

### Affected Resource(s)

terraform-oci-oke  >  module.oke.null_resource.install_kubectl_operator

### Terraform Configuration Files

https://gist.github.com/bieksaz/c2076e7602a1d9eceef43aced4fe2bb9#file-oci_oke_kubectl_config

### Normal Output (no debug captured)

https://gist.github.com/bieksaz/03025ae359486fed58b5b943527788c7#file-oci_oke_kubectl_random_failure_log

### Expected Behavior

Should not time out (it does work from time to time)

### Actual Behavior

on random deployments it fails:

module.oke.null_resource.install_kubectl_operator[0] (remote-exec): ol7_olcne        | 2.8 kB     00:00
module.oke.null_resource.install_kubectl_operator[0] (remote-exec): ol7_olcne11      | 2.8 kB     00:00
module.oke.null_resource.install_kubectl_operator[0] (remote-exec): (2/4): ol7_olcn 0% |    0 B   --:-- ETA
module.oke.null_resource.install_kubectl_operator[0] (remote-exec): (1/4): ol7_olcne/x |  62 kB   00:00
module.oke.null_resource.install_kubectl_operator[0] (remote-exec): (2/4): ol7_olcne11 |  37 kB   00:00
module.oke.null_resource.install_kubectl_operator[0] (remote-exec): (3/4): ol7_olcne/x |  33 kB   00:00
module.oke.null_resource.install_kubectl_operator[0] (remote-exec): (4/4): ol7_olcne11 | 163 kB   00:00
module.oke.null_resource.install_kubectl_operator[0] (remote-exec):  One of the configured repositories failed (Unknown),
module.oke.null_resource.install_kubectl_operator[0] (remote-exec):  and yum doesn't have enough cached data to continue. At this point the only
.....
module.oke.null_resource.install_kubectl_operator[0] (remote-exec): ('Connection aborted.', error(111, 'Connection refused'))

### Steps to Reproduce

redeploy

**Comment by karthicgit** [comment]
@bieksaz  May  I know  you are adding the created operator instance into OS management?

**Comment by bieksaz** [comment]
Hi, I could not understand what exactly are you asking for?
p.s. its a regular  terraform-oci-oke (v2.3.2) module that we are running and it has operator deployment in it. Starting from v2.3.2 operator has also kubectl installed. I am sure you know it.

**Comment by karthicgit** [comment]
@bieksaz When its failing could you please check the yum repo in operator instance --> /etc/yum.repos.d/oracle-olcne-ol7.repo and look for 

[ol7_olcne11]
name=Oracle Linux Cloud Native Environment version 1.1 ($basearch)
baseurl=https://yum$ociregion.oracle.com/repo/OracleLinux/OL7/olcne11/$basearch/
gpgkey=file:///etc/pki/rpm-gpg/RPM-GPG-KEY-oracle
gpgcheck=1
enabled=1
Then try to run the command "sudo yum install -y kubectl". You mentioned its failing intermittently is there any proxy involved?

**Comment by bieksaz** [comment]
HI, just installed with a failure. Here's what we have there:
File /etc/yum.repos.d/oracle-olcne-ol7.repo does not exist, only  /etc/yum.repos.d/oracle-olcne-ol7.repo.osms-backup is there

[opc@qh-operator ~]$ ls -la /etc/yum.repos.d/
total 56
drwxr-xr-x.  2 root root 4096 Oct 28 08:50 .
drwxr-xr-x. 95 root root 8192 Oct 28 08:59 ..
-rw-r--r--.  1 root root  488 Jun 11 22:19 ksplice-ol7.repo.osms-backup
-rw-r--r--.  1 root root  205 Oct 28 08:49 ksplice-uptrack.repo.osms-backup
-rw-r--r--.  1 root root  252 Jun 11 22:22 oci-included-ol7.repo.osms-backup
-rw-r--r--.  1 root root  252 Jun 11 22:19 oracle-epel-ol7.repo.osms-backup
-rw-r--r--.  1 root root 1039 Dec 13  2019 oraclelinux-developer-ol7.repo.osms-backup
-rw-r--r--.  1 root root 4082 Sep 17 01:29 oracle-linux-ol7.repo.osms-backup
-rw-r--r--.  1 root root  723 Oct 28 08:50 oracle-olcne-ol7.repo.osms-backup
-rw-r--r--.  1 root root  276 Jun 11 22:25 oracle-softwarecollection-ol7.repo.osms-backup
-rw-r--r--.  1 root root 2587 Jul  8 22:25 uek-ol7.repo.osms-backup
-rw-r--r--.  1 root root  226 Jul  8 22:25 virt-ol7.repo.osms-backup


Contents of /etc/yum.repos.d/oracle-olcne-ol7.repo.osms-backup

[opc@qh-operator ~]$ cat /etc/yum.repos.d/oracle-olcne-ol7.repo.osms-backup
[ol7_olcne]
name=Oracle Linux Cloud Native Environment version 1.0 ($basearch)
baseurl=https://yum$ociregion.oracle.com/repo/OracleLinux/OL7/olcne/$basearch/
gpgkey=file:///etc/pki/rpm-gpg/RPM-GPG-KEY-oracle
gpgcheck=1
enabled=1

[ol7_olcne11]
name=Oracle Linux Cloud Native Environment version 1.1 ($basearch)
baseurl=https://yum$ociregion.oracle.com/repo/OracleLinux/OL7/olcne11/$basearch/
gpgkey=file:///etc/pki/rpm-gpg/RPM-GPG-KEY-oracle
gpgcheck=1
enabled=1

[ol7_developer_olcne]
name=Developer Preview for Oracle Linux Cloud Native Environment ($basearch)
baseurl=https://yum$ociregion.oracle.com/repo/OracleLinux/OL7/developer/olcne/$basearch/
gpgkey=file:///etc/pki/rpm-gpg/RPM-GPG-KEY-oracle
gpgcheck=1
enabled=0




CURL output:

[opc@qh-operator ~]$ curl -v https://yum-london.oracle.com/repo/OracleLinux/OL7/olcne11/x86_64/
* About to connect() to yum-london.oracle.com port 443 (#0)
*   Trying 130.35.114.76...
* Connected to yum-london.oracle.com (130.35.114.76) port 443 (#0)
* Initializing NSS with certpath: sql:/etc/pki/nssdb
*   CAfile: /etc/pki/tls/certs/ca-bundle.crt
  CApath: none
* SSL connection using TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
* Server certificate:
*       subject: CN=yum-uk-london-1.oracle.com,OU=Oracle CORPARCH LONDON,O=Oracle Corporation,L=Redwood City,ST=California,C=US
*       start date: Feb 20 00:00:00 2019 GMT
*       expire date: Feb 19 12:00:00 2021 GMT
*       common name: yum-uk-london-1.oracle.com
*       issuer: CN=DigiCert SHA2 Secure Server CA,O=DigiCert Inc,C=US
> GET /repo/OracleLinux/OL7/olcne11/x86_64/ HTTP/1.1
> User-Agent: curl/7.29.0
> Host: yum-london.oracle.com
> Accept: */*
>
< HTTP/1.1 200 OK
< Server: nginx
< Date: Wed, 28 Oct 2020 10:17:55 GMT
< Content-Type: text/html
< Transfer-Encoding: chunked
< Connection: keep-alive
<
<html>
<head><title>Index of /repo/OracleLinux/OL7/olcne11/x86_64/</title></head>
<body bgcolor="white">
<h1>Index of /repo/OracleLinux/OL7/olcne11/x86_64/</h1><hr><pre><a href="../">../</a>
<a href="getPackage/">getPackage/</a>                                        21-Oct-2020 08:28                   -
<a href="getPackageSource/">getPackageSource/</a>                                  21-Oct-2020 08:27                   -
<a href="repodata/">repodata/</a>                                          28-Oct-2020 04:10                   -
<a href="00691ac4f72cdebd6f758bed367608f51852c801-updateinfo.xml.gz">00691ac4f72cdebd6f758bed367608f51852c801-update..&gt;</a> 01-May-2020 04:04                6006
<a href="04aa1aba842365ce774b952db47b954f578ef44d-updateinfo.xml.gz">04aa1aba842365ce774b952db47b954f578ef44d-update..&gt;</a> 22-Jul-2020 19:45                7817
<a href="0dcfd4b646893db8958c80eec8ca87c784b8d60c-updateinfo.xml.gz">0dcfd4b646893db8958c80eec8ca87c784b8d60c-update..&gt;</a> 14-Sep-2020 11:45               15437
<a href="18d2381e6627d6e7a22067663c8315a401b2e653-updateinfo.xml.gz">18d2381e6627d6e7a22067663c8315a401b2e653-update..&gt;</a> 24-Sep-2020 15:46               21616
<a href="54954fc0f011410ee63a9acb85cdf1bcb6208cc4-updateinfo.xml.gz">54954fc0f011410ee63a9acb85cdf1bcb6208cc4-update..&gt;</a> 28-Oct-2020 04:10               37735
<a href="5f7d8698dd623bde2fc797f69f408dcb122222ea-updateinfo.xml.gz">5f7d8698dd623bde2fc797f69f408dcb122222ea-update..&gt;</a> 13-Oct-2020 15:04               27404
<a href="7d9fda893551b049c39d3db0955bd4731a727045-updateinfo.xml.gz">7d9fda893551b049c39d3db0955bd4731a727045-update..&gt;</a> 12-Jun-2020 15:43                6478
<a href="a71de87c7c111197c16bad7ef7c9c313c628612b-updateinfo.xml.gz">a71de87c7c111197c16bad7ef7c9c313c628612b-update..&gt;</a> 31-Aug-2020 11:43               12344
<a href="b7752cfaa600571dd4b7bb5ea6d6232a8f1c614f-updateinfo.xml.gz">b7752cfaa600571dd4b7bb5ea6d6232a8f1c614f-update..&gt;</a> 17-Aug-2020 10:40               11233
<a href="ddc22dafcce607ad57efb21cf7dfbcdd3f45e2f0-updateinfo.xml.gz">ddc22dafcce607ad57efb21cf7dfbcdd3f45e2f0-update..&gt;</a> 08-Aug-2020 23:42               10526
<a href="e69dc23001d67e3523ab7a11297de4a2817d77ee-updateinfo.xml.gz">e69dc23001d67e3523ab7a11297de4a2817d77ee-update..&gt;</a> 20-Oct-2020 03:58               43383
</pre><hr></body>
</html>
* Connection #0 to host yum-london.oracle.com left intact



sudo yum install -y kubectl     als oworked fine when executed manually.


There is no proxy, nothing. Its my private computer connected directly to internet

**Comment by hyder** [comment]
Ah it's the OSMS that's messing with the yum repos on the operator host.

**Comment by hyder** [comment]
@bieksaz: can you please test:

```
git checkout -b hyder-issue-225 master
git pull https://github.com/hyder/terraform-oci-oke.git issue-225
git checkout master
git merge --no-ff hyder-issue-225
git push origin master
```

You'll need to re-init your terraform-oci-oke module.

Let us know how you go.

**Comment by bieksaz** [comment]
Hi, I'll do this tomorrow morning, its late already. where should I run this? On Operator host? On my own laptop? On my docker image from where I run terraform deployment scripts? 
And its not entirely clear what should I checkout (first statement)? I should pull some repo first probably from git? Could it be 1 and 2 lines should be swapped?   actually not really following this whole test scenario.   Firstly I checkout some  branch of unknown repo, then pull oci-oke repo inside first unknown repo as subrepo (or how is it)?  then checkout master of this unknown repo then merge without commit?   what should happen here?   not really following. Can you perhaps explain a bit more in detail what you need me to do?

**Comment by hyder** [comment]
@bieksaz you should do this from the location where you've cloned the terraform scripts. 

Sorry, I copied the merge commands from github. I've fixed it now.

The first statement will create a local branch. You'll then make a pull from my repo where I've created a patch that will disable the OSMS on the operator host. Run terraform init and recreate the whole infrastructure/run your whole test. 

Basically, we want to ensure that installation of kubectl won't fail.

This will go in the next release if your test is successful.

**Comment by bieksaz** [comment]
I am not using  terraform-oci-oke git repo anymore, we downloaded it locally and now using stale files from  terraform-oci-oke 
 branch v2.3.2.  So my understanding is: you want me to test installation using terraform-oci-oke branch issue-225, correct? if so, I will replace current version with the one that's in issue-225 and run a test

**Comment by hyder** [comment]
Yes. However, please note that branch issue-225 is under my fork. The only diff is that it uses a new version of base which itself has a new version for operator where osms is disabled.

**Comment by bieksaz** [comment]
Hi, @hyder 

I tried to replace terraform-oci-oke with your branch issue-225 and here's what I got on plan


Error: Invalid function argument

  on modules/oke/datasources.tf line 9, in data "oci_core_images" "latest_images":
   9:   shape                    = lookup(each.value,"shape","VM.Standard.E3.Flex")
    |----------------
    | each.value is tuple with 2 elements

Invalid value for "inputMap" parameter: lookup() requires a map as the first
argument.
 


I see you're passing 4 values for nodepool: 
np1 = {shape="VM.Standard.E3.Flex",ocpus=2,node_pool_size=2,boot_volume_size=150}  
and then you're searching for "shape"  "VM.Standard.E3.Flex"   in datasources.tf. It is hardcoded value, which does not look right..  

In my case on branch v2.3.2 I was only passing over 2 values and without value names like "shape", which worked fine in past. Just like this:   np1 = ["VM.Standard.E2.1", 3]

Now i made two tests: with your defaults ("VM.Standard.E3.Flex") and with "VM.Standard.E2.1" after changing same value in datasources.tf.  Both times deployed successfully. That indicates your fix does solve this random failure issue, but this solution is not entirely good, as it has a hardcoded value in datasources.tf  and that cant be committed this way.

**Comment by hyder** [comment]
Thanks for testing and the feedback. Yes, I see the problem. 

So, we've recently added support for the flex shape and it's already in master branch but it's only available as a release candidate. However, when I created branch issue-225 for this fix, I created the branch off master. My apologies for that.

Note that the hardcoded value in datasources is just the default value in case a shape is not specified (see https://www.terraform.io/docs/configuration/functions/lookup.html) but it does expect a number of other parameters from your var file. 

I'll create a separate release for the 2.3.x line. 

Thanks again for testing.

**Comment by hyder** [comment]
The good thing is that we now know turning OSMS off, albeit temporarily, fixes your current issue with kubectl as well as @karthicgit issue with helm in #212

**Comment by hyder** [comment]
Hi @bieksaz ,

Can you please try branch patch-225? https://github.com/oracle-terraform-modules/terraform-oci-oke/tree/patch-225

If this works, we'll cut a new release for the 2.x line.

**Comment by bieksaz** [comment]
hi @hyder,
tested your new branch with  original NP parameters I used in past and it worked:  np1 = ["VM.Standard.E2.1", 3]

do yea, please push this fix and I add it in my codebase locally

**Comment by hyder** [comment]
thanks for testing. I'll maintain it in this branch and cut a new release. You can use the new release.

**Comment by hyder** [comment]
Here you go: https://github.com/oracle-terraform-modules/terraform-oci-oke/releases/tag/v2.3.3

Thanks very much for testing and your patience. Very much appreciated.
